api/catalog/api/models/audio.py

```diff
@@ -141,6 +142,14 @@ class Audio(AudioFileMixin, AbstractMedia):
     category: eg. music, sound_effect, podcast, news & audiobook
     """
 
+    audioset = models.ForeignObject(
+        to="AudioSet",
+        on_delete=models.DO_NOTHING,
```

Comment:
What is the intention behind using "do nothing" for this?

Reply:
We don't want the audio hard-linked to audio sets. Both tables are destroyed and recreated during ingestion so
- we don't want the deletes to cascade
- we cannot maintain DB integrity during ingestion

Also, this behaviour is identical to the one we had before, so I went for minimal disruption.

Reply:
It to me a second to understand how `on_delete` might interact with `db_constraint=False`. I was trying to square it with the Postgres behaviour, where the constraint enforcement is delayed until the end of the transaction in the "do nothing" setting, but I suppose that Django does not enforce it "manually" for us, does it?

Reply:
I'm sorry I didn't understand your question. But I'll take a stab to answer anyway.

1. `on_delete="DO_NOTHING"` prevents the model itself from being deleted or the field value from being changed when the referenced object goes away.
2. `db_contraint=False` allows the field to accept values that are not present in the referenced table.

These are both for situations that could happen during ingestion as both `audio` and `audioset` tables are recreated and filled separately.

Reply:
Right, that is my understanding of the current implementation as well, but critically it's the usage db_constraint=False in conjunction with DO_NOTHING that results in that. In regular db-enforced FK constraints in Postgres, the "NO ACTION" delete effect still enforces the constraint on a DB level, it just delays the check until the end of the transaction instead of immediately: https://www.postgresql.org/docs/15/ddl-constraints.html#DDL-CONSTRAINTS-EXCLUSION

> Restricting and cascading deletes are the two most common options. RESTRICT prevents deletion of a referenced row. NO ACTION means that if any referencing rows still exist when the constraint is checked, an error is raised; this is the default behavior if you do not specify anything. (**The essential difference between these two choices is that NO ACTION allows the check to be deferred until later in the transaction, whereas RESTRICT does not.**) CASCADE specifies that when a referenced row is deleted, row(s) referencing it should be automatically deleted as well. There are two other options: SET NULL and SET DEFAULT. These cause the referencing column(s) in the referencing row(s) to be set to nulls or their default values, respectively, when the referenced row is deleted. Note that these do not excuse you from observing any constraints. For example, if an action specifies SET DEFAULT but the default value would not satisfy the foreign key constraint, the operation will fail.

I've added bold emphasis to the relevant section of the quoted paragraph.

Critically: "no action" (or "do nothing" as Django calls it) _still enforces the constraint_, just not on the DB level. I'm _assuming_ that Django doesn't enforce it programmatically because [the documentation for `db_constraint`](https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.ForeignKey.db_constraint) doesn't mention it doing anything of that nature and even that it's "very bad for data integrity" which seems to imply that using this setting makes it easy to get into a bad state (which we already know anyway :slightly_smiling_face:)

In any case, I was mostly just thinking out loud to understand how "do nothing" and db_constraint=False were working together to make it so that we can freely delete media items without causing damage to media reports, waveforms, and other related data.